lib/ZnapZend/ZFS.pm

```diff
@@ -431,28 +456,104 @@ sub sendRecvSnapshots {
 sub getDataSetProperties {
     my $self = shift;
     my $dataSet = shift;
+    my $recurse = shift; # May be not passed => undef
+
     my @propertyList;
     my $propertyPrefix = $self->propertyPrefix;
 
-    my $list = $dataSet ? [ ($dataSet) ] : $self->listDataSets();
+    my @list;
+    print STDERR "=== getDataSetProperties():\n\trecurse=" . Dumper($recurse) . "\n\tDS=" . Dumper($dataSet) . "\n\tlowmemRecurse=" . $self->lowmemRecurse . "\n" if $self->debug;
+
+    if (!defined($recurse)) {
+        $recurse = 0;
+    }
+
+    # Note: Before the recursive and multiple dataset support we either
+    # used the provided dataSet name "as is" trusting it exists (failed
+    # later if not), or called listDataSets() to list everything on the
+    # system from `zfs` (also ensuring stuff exists). So we still do.
+    # Before the recursive the @list must have had individual dataset
+    # names, passed directly, or subsequently recursively listed above,
+    # to assign into the discovered list elements. So no recursion was
+    # needed in the logic below. Now we recurse by default to call `zfs`
+    # as rarely as we can and so complete faster. However, on systems
+    # with too many datasets this can exhaust the memory available to
+    # the process, and/or time out. To defend against this, we optionally
+    # can fall back to a big list of individual dataset names found by
+    # recursive listDataSets() invocations instead.
+    if ($dataSet) {
+        if (ref($dataSet) eq 'ARRAY') {
+            print STDERR "=== getDataSetProperties(): Is array...\n" if $self->debug;
+            if ($self->lowmemRecurse && $recurse) {
+                my $listds = $self->listDataSets(undef, $dataSet, $recurse);
```

Comment:
Also a question for reviewers : in the fallback logic here and a bit below, I pass `undef` for the `$remote` meaning to look at the local system. First it seemed to match the earlier call when used without any arguments, or verbatim use of the `$dataSet` string when with arguments.

However here we can pass several args (as an array), some of which potentially could have the `user@remote:data/set` format... should I bother about this possibility now? :-) I do not think this change breaks any use-case fundamentally, but maybe others have more visibility and memory :)

Reply:
I would suggest to have as little ssh magic on the commandline and rather encurrage people to create appropriate .ssh/config files and then work with `remote` names matching these definitions.